app/services/order_syncer.rb

```diff
@@ -66,14 +66,13 @@ def update_payment_for(order)
   end
 
   def update_shipment_for(order)
-    shipment = order.shipments.with_state('pending').where(shipping_method_id: shipping_method_id_was).last
-    if shipment
-      shipment.update_attributes(shipping_method_id: shipping_method_id)
-      order.update_attribute(:shipping_method_id, shipping_method_id)
-    else
-      unless order.shipments.with_state('pending').where(shipping_method_id: shipping_method_id).any?
-        order_update_issues.add(order, I18n.t('admin.shipping_method'))
-      end
+    shipment = order.shipment
+    return track_shipping_method_issue(order) if shipment.blank?
+
+    if shipment.state == "pending" && shipment.shipping_method.id == shipping_method_id_was
+      order.select_shipping_method(shipping_method_id)
+    elsif !(shipment.state == "pending" && shipment.shipping_method.id == shipping_method_id)
+      track_shipping_method_issue(order)
```

Comment:
this is too difficult to read,
we should simplify this, even if only extracted to methods with good names.

or maybe just remove the elsif,
if we just have two ifs one after the other, it's more readable and more correct, in case the first assignment fails, right?

```ruby
    if shipment.state == "pending" && shipment.shipping_method.id == shipping_method_id_was
      order.select_shipping_method(shipping_method_id)
    end
    unless shipment.state == "pending" && shipment.shipping_method.id == shipping_method_id
      track_shipping_method_issue(order)
     end
```
and then, is this correct?
```ruby
    track_shipping_method_issue(order) unless shipment.state == "pending"
    if shipment.shipping_method.id == shipping_method_id_was
      order.select_shipping_method(shipping_method_id)
    end
    track_shipping_method_issue(order) unless shipment.shipping_method.id == shipping_method_id
```

Reply:
The first snippet works is equivalent with a `return` in the first block. :+1: 

But the next version is not equivalent. The current version does not always call `track_shipping_method_issue(order)` if shipment state is not pending, but this would.

Thinking about this...

Reply:
I added a commit. It's a lot more readable now: https://github.com/openfoodfoundation/openfoodnetwork/pull/3560/commits/7505dd0410043d51065122c7db65c2032ce32226

Thanks, @luisramos0! :slightly_smiling_face:

Reply:
ah, yes, awesome, much better!